File: SpringProject/src/main/java/org/example/springproject/service/UserService.java
/**
 * UserService.java
 * This interface defines the contract for user-related operations in the application.
 * @author Ilea Robert-Ioan
 */
package org.example.springproject.service;

import org.example.springproject.dto.UserDTO;
import org.example.springproject.entity.User;

import java.util.List;

/**
 * UserService interface provides methods for managing users in the application.
 * It includes operations for adding, deleting, updating, and retrieving user information.
 */
public interface UserService {

    /**
     * Adds a new user to the system.
     * @param user The user to be added.
     * @return UserDTO containing the details of the added user.
     * @throws RuntimeException if an error occurs during the addition of the user.
     */
    UserDTO addUser(User user);

    /**
     * Deletes a user by their ID.
     * @param id The ID of the user to be deleted.
     * @return UserDTO containing the details of the deleted user.
     * @throws RuntimeException if an error occurs during the deletion of the user.
     */
    UserDTO deleteUserById(String id);

    /**
     * Updates an existing user.
     * @param id The ID of the user to be updated.
     * @param updatedUser The updated user information.
     * @return UserDTO containing the details of the updated user.
     */
    UserDTO updateUser(String id, User updatedUser);

    /**
     * Retrieves a list of all users.
     * @return List of UserDTO containing details of all users.
     */
    List<UserDTO> getUsers();

    /**
     * Retrieves a user by their ID.
     * @param id The ID of the user to be retrieved.
     * @return UserDTO containing the details of the user.
     */
    UserDTO getUserById(String id);

    /**
     * Retrieves a user by their email address.
     * @param email The email address of the user to be retrieved.
     * @return User containing the details of the user.
     */
    User getUserByEmail(String email);

    /**
     * Retrieves a user ID by their email address.
     * @param email The email address of the user.
     * @return String containing the user ID.
     */
    String getUserIdByEmail(String email);

    /**
     * Retrieves a user by their room ID.
     * @param roomId The room ID associated with the user.
     * @return UserDTO containing the details of the user.
     */
    UserDTO getUserByRoomId(String roomId);

    /**
     * Updates the phone number of a user.
     * @param userId The ID of the user whose phone number is to be updated.
     * @param phone The new phone number to be set for the user.
     */
    void updateUserPhone(String userId, String phone);

    /**
     * Retrieves the phone number of a user by their user ID.
     * @param userId The ID of the user whose phone number is to be retrieved.
     * @return String containing the user's phone number.
     */
    String getUserPhoneNumber(String userId);
}

File: AngularProject/src/app/services/user.service.ts
/**
 * user.service.ts
 * This service handles user-related operations.
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { User } from '../models/user-model';
const BASE_URL = 'http://localhost:8080';

@Injectable({
  providedIn: 'root'
})
/**
 * UserService class
 * This service provides methods to interact with user-related endpoints.
 */
export class UserService {

  /**
   * Constructor for UserService
   * @param http - HttpClient instance for making HTTP requests
   */
  constructor(private http: HttpClient) { }

  /**
   * Get user ID by email
   * @param email - User's email address
   */
  get_user_id_by_email(email: string):Observable<string>{
    return this.http.get(`${BASE_URL}/user/email/${email}`,{
          responseType: 'text'
    });
  }

  /**
   * Get user phone number by user ID
   * @param userId - User's ID
   */
  get_user_phone_by_id(userId: string): Observable<string> {
    return this.http.get(`${BASE_URL}/user/phone`, {
      params: { userId },
      responseType: 'text'
    });
  }

  /**
   * Get user details by user ID
   * @param userId - User's ID
   */
  get_user_by_id(userId: string): Observable<User> {
    return this.http.get<User>(`${BASE_URL}/user/${userId}`);
  }

  /**
   * Update user phone number
   * @param userId - User's ID
   * @param phone - New phone number
   */
  update_user_phone(userId: string, phone: string): Observable<string> {
    return this.http.put(`${BASE_URL}/user/${userId}/phone`, null, {
      params: { phone },
      responseType: 'text'
    });
  }

}
